internal/jobs/jobobject.go

```diff
@@ -0,0 +1,363 @@
+// for notifications and then returns an object with the corresponding handles.
+func CreateJobObject(name string) (*JobObject, error) {
+	jobHandle, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr(name))
+	if err != nil {
+		return nil, err
+	}
+
+	iocpChan := make(chan IOCPNotif, 10)
+	jobMap[uint32(jobHandle)] = iocpChan
+	// If the IOCP we'll be using to receive messages for all jobs hasn't been
+	// created, create it and start polling.
+	if ioCompletionPort == 0 {
+		ioInitOnce.Do(initIo)
+	}
+
+	if _, err = attachIOCP(jobHandle, ioCompletionPort); err != nil {
+		windows.Close(jobHandle)
+		return nil, err
+	}
+	return &JobObject{
+		jobHandle,
+		iocpChan,
+		sync.RWMutex{},
+	}, nil
+}
+
+// IOCPChan returns the channel that the job object will receive IO completion port
+// notifications on.
+func (job *JobObject) IOCPChan() chan IOCPNotif {
+	return job.iocpChan
+}
+
+// Close closes the job object and iocp handles. If this is the only open handle
+// the job object will be terminated.
+func (job *JobObject) Close() error {
+	if job.jobHandle == 0 {
+		return nil
+	}
+	job.handleLock.Lock()
+	defer job.handleLock.Unlock()
+
+	if job.jobHandle != 0 {
+		if err := windows.Close(job.jobHandle); err != nil {
+			return err
+		}
+		job.jobHandle = 0
+	}
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *JobObject) Assign(pid uint32) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if pid == 0 {
+		return errors.New("process has not started")
+	}
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, pid)
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.jobHandle, hProc)
+}
+
+// Terminate terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *JobObject) Terminate() error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+	return windows.TerminateJobObject(job.jobHandle, 1)
+}
+
+// Shutdown gracefully shuts down all the processes in the job object.
+func (job *JobObject) Shutdown(ctx context.Context) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	var (
+		terminate bool
+		signalErr bool
+	)
+	pids, err := job.Pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Errorf("failed to send ctrl-break to process in job: %s", err)
+			signalErr = true
+		}
+	}
+
+	// Get pids in job again to check if they were successfully killed by the signals.
+	// Certain programs (ping.exe im looking at you) handle ctrl signals so we need
```

Comment:
lol but we should probably take the "im looking at you" part out :P

Reply:
Aww I kinda wanted this to stay haha

Reply:
Yeah it's funny but it's not good comment/code style :/

Reply:
I'll remove